File: src/types/vector.ts
export class Vector3 {
    x: number;
    y: number;
    z: number;

    constructor(x: number, y: number, z: number) {
        this.x = x;
        this.y = y;
        this.z = z;
    }

    toGLSL() {
        return `vec3(${this.x},${this.y},${this.z})`;
    }

    toArray() {
        return [this.x, this.y, this.z];
    }

    add(other: Vector3): Vector3 {
        return new Vector3(this.x + other.x, this.y + other.y, this.z + other.z);
    }
}

File: src-tauri/src/vector.rs
use std::ops::Add;

use serde::{Serialize, Deserialize};
use ts_rs::TS;


#[derive(Debug, Serialize, Deserialize, Clone, TS)]
#[ts(export, export_to = "../src/bindings/")]
pub struct Vector3 {
    pub x: f32,
    pub y: f32,
    pub z: f32,
}

impl Vector3 {
    pub fn to_glsl(self) -> String {
        return format!("vec3({:.10},{:.10},{:.10})", self.x, self.y, self.z);
    }
}

impl Add for Vector3 {
    type Output = Vector3;

    fn add(self, other: Vector3) -> Vector3 {
        Vector3 {
            x: self.x + other.x,
            y: self.y + other.y,
            z: self.z + other.z,
        }
    }
}